{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fonction qui énumère les différent kmers possibles\n",
    "\n",
    "def enumerateWords(seq,length):\n",
    "    numberOfWords = len(seq)-length+1\n",
    "    words = []\n",
    "    for i in range(numberOfWords):\n",
    "        words.append(seq[i:i+length])\n",
    "    return words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fonction qui trouve les matches entre un kmer et un mot\n",
    "\n",
    "def findMatches(kmers,word,seed):\n",
    "    matches = []\n",
    "    for i in range(len(kmers)):\n",
    "        for j in range(len(word)):\n",
    "            if((seed[j] == '1') and (word[j] != kmers[i][j])):\n",
    "                break\n",
    "            if(j == len(word)-1):\n",
    "                matches.append(i)\n",
    "    return matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fonction qui étend les HSP et retourne les nouveaux indices et le score\n",
    "\n",
    "def extendHsp(seq1,i,seq2,j,seed,e):\n",
    "    \n",
    "    #calculer le score initial\n",
    "    nbOnes = 0\n",
    "    for k in range(len(seed)):\n",
    "        if(seed[k] == \"1\"):\n",
    "            nbOnes += 1\n",
    "    \n",
    "    score = nbOnes*5 + (len(seed)-nbOnes)*-4\n",
    "    scoreToReturn = score\n",
    "    scoreMax = score\n",
    "    \n",
    "    #mettre les indices de début et de fin dans des variables\n",
    "    start1 = i\n",
    "    end1 = i+len(seed)-1\n",
    "    start2 = j\n",
    "    end2 = j+len(seed)-1\n",
    "    \n",
    "    #vérifier si on est au début où la fin du séquence\n",
    "    canExtendLeft = True if (start1 > 0 and start2 > 0) else False\n",
    "    canExtendRight = True if (end1 < len(seq1)-1 and end2 < len(seq2)-1) else False\n",
    "    \n",
    "    #boucler tant qu'on peut étendre et que le score ne baisse pas sous le seuil\n",
    "    while((canExtendLeft or canExtendRight) and (score > scoreMax-e)):\n",
    "        scoreLeft = -float(\"Inf\")\n",
    "        scoreRight = -float(\"Inf\")\n",
    "        scoreBoth = -float(\"Inf\")\n",
    "        \n",
    "        #calculer les trois scores possibles\n",
    "        if(canExtendLeft):\n",
    "            scoreLeft = score + (5 if (seq1[start1-1] == seq2[start2-1]) else -4)\n",
    "            \n",
    "        if(canExtendRight):\n",
    "            scoreRight = score + (5 if (seq1[end1+1] == seq2[end2+1]) else -4)\n",
    "            \n",
    "        if(canExtendRight and canExtendLeft):\n",
    "            scoreBoth = scoreLeft + scoreRight -score\n",
    "            \n",
    "        #trouver le meilleur score\n",
    "        scores = [scoreLeft,scoreRight,scoreBoth]\n",
    "        indexScoreMax = np.argmax(np.array(scores))\n",
    "        score = max(scoreLeft,scoreRight,scoreBoth)\n",
    "        \n",
    "        if(score > scoreMax):\n",
    "            scoreMax = score\n",
    "        \n",
    "        #agrandir le HSP si possible\n",
    "        if(score > scoreMax-e):\n",
    "            scoreToReturn = score\n",
    "            if(indexScoreMax == 0 or indexScoreMax == 2):\n",
    "                start1 -= 1\n",
    "                start2 -= 1\n",
    "            if(indexScoreMax == 1 or indexScoreMax == 2):\n",
    "                end1 += 1\n",
    "                end2 += 1\n",
    "            \n",
    "        #recalculer la possibilité d'extension\n",
    "        canExtendLeft = True if (start1 > 0 and start2 > 0) else False\n",
    "        canExtendRight = True if (end1 < len(seq1)-1 and end2 < len(seq2)-1) else False\n",
    "        \n",
    "    return (start1,end1,start2,end2,scoreToReturn)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calcul du bitscore\n",
    "\n",
    "def computeBitscore(score):\n",
    "    return round((0.192*score-math.log(0.176))/math.log(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calcul de la e-value\n",
    "\n",
    "def computeEvalue(m,n,b):\n",
    "    return m*n*(2**-b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fonction qui retourne le score pour deux sous-séquences\n",
    "\n",
    "def computeScore(seq1,start1,end1,seq2,start2,end2):\n",
    "    length = end1 - start1 + 1\n",
    "    score = 0\n",
    "    for i in range(length):\n",
    "        if(seq1[start1+i] == seq2[start2+i]):\n",
    "            score += 5\n",
    "        else:\n",
    "            score -= 4\n",
    "            \n",
    "    return score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fonction retournant une liste de tous les HSP optimalement fusionnés\n",
    "\n",
    "def mergeHsps(hsps,seq1,seq2):\n",
    "    \n",
    "    merged_hsps_new = hsps\n",
    "    merged_hsps = []\n",
    "    \n",
    "    #continuer à essayer de fusionner les HSP tant que la dernière et l'avant-dernière liste\n",
    "    #n'ont pas la même taille (tous fusionnés)\n",
    "    while(len(merged_hsps) != len(merged_hsps_new)):\n",
    "        merged_hsps = merged_hsps_new\n",
    "        merged_hsps_new = []\n",
    "        \n",
    "        #garder une liste des HSP fusionnés pour ne pas les réinclure dans la liste\n",
    "        hasMerged = np.zeros((len(merged_hsps)))\n",
    "        \n",
    "        #regarder chaque paire de HSP\n",
    "        for i in range(len(merged_hsps)-1):\n",
    "            for j in range(i+1,len(merged_hsps)):\n",
    "                #indices du HSP de la 1ere séquence\n",
    "                start1_1 = merged_hsps[i][0]\n",
    "                end1_1 = merged_hsps[i][1]\n",
    "                start2_1 = merged_hsps[i][2]\n",
    "                end2_1 = merged_hsps[i][3]\n",
    "                \n",
    "                #2e séquence, mutatis mutandis\n",
    "                start1_2 = merged_hsps[j][0]\n",
    "                end1_2 = merged_hsps[j][1]\n",
    "                start2_2 = merged_hsps[j][2]\n",
    "                end2_2 = merged_hsps[j][3]\n",
    "                                \n",
    "                #cas où la 1ere commence avant la 2e\n",
    "                if(start1_1 < start1_2 and end1_1 > start1_2):\n",
    "                    overlap = end1_1 - start1_2\n",
    "                    #vérifier s'il y a vraiment overlap\n",
    "                    if(end2_1 - overlap  == start2_2):\n",
    "                        hasMerged[i] = 1\n",
    "                        hasMerged[j] = 1\n",
    "                        score = computeScore(seq1,start1_1,end1_2,seq2,start2_1,end2_2)\n",
    "                        newHsp = (start1_1, end1_2, start2_1, end2_2, score)\n",
    "                        #ajouter la fusion à la nouvelle liste\n",
    "                        merged_hsps_new.append(newHsp)\n",
    "                #cas inverse\n",
    "                elif(start1_1 > start1_2 and end1_2 > start1_1):\n",
    "                    overlap = end1_2 - start1_1\n",
    "                    if(end2_2 - overlap == start2_1):\n",
    "                        hasMerged[i] = 1\n",
    "                        hasMerged[j] = 1\n",
    "                        score = computeScore(seq1,start1_2,end1_1,seq2,start2_2,end2_1)\n",
    "                        newHsp = (start1_2, end1_1, start2_2, end2_1, score)\n",
    "                        merged_hsps_new.append(newHsp)\n",
    "        \n",
    "        #ajouter les HSP non fusionnés à la nouvelle liste\n",
    "        for k in range(len(merged_hsps)):\n",
    "            if(hasMerged[k] == 0):\n",
    "                merged_hsps_new.append(merged_hsps[k])\n",
    "                \n",
    "    return merged_hsps_new\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fonction qui retourne les indices (pour les deux séquences) et le score du meilleur HSP\n",
    "\n",
    "def getBestHsp(seq1,seq2,e,ss,seed):\n",
    "    #séquence à chercher \n",
    "    kmersSearched = enumerateWords(seq1,len(seed))\n",
    "    #séquence dans la BD\n",
    "    kmersSearching = enumerateWords(seq2,len(seed))\n",
    "    \n",
    "    #trouver les kmers qui match selon la graine\n",
    "    hsps = []\n",
    "    for i in range(len(kmersSearched)):\n",
    "        matches = findMatches(kmersSearching,kmersSearched[i],seed)\n",
    "        \n",
    "        if(matches != []):\n",
    "            for j in range(len(matches)):\n",
    "                hsps.append((matches[j],i))\n",
    "         \n",
    "    #retourner None si on ne trouve aucun HSP\n",
    "    if(len(hsps) == 0):\n",
    "        return None\n",
    "                \n",
    "    #étendre les HSP     \n",
    "    extended_hsps = []\n",
    "    for k in range(len(hsps)):\n",
    "        extended_hsps.append(extendHsp(seq2,hsps[k][0],seq1,hsps[k][1],seed,e))\n",
    "        \n",
    "    #supprimer les doublons        \n",
    "    extended_hsps_cleaned = list(dict.fromkeys(extended_hsps))\n",
    "    \n",
    "    #merger les HSP\n",
    "    merged_hsps = mergeHsps(extended_hsps_cleaned,seq2,seq1)\n",
    "\n",
    "    #trouver le meilleur HSP\n",
    "    best = 0\n",
    "    scoreMax = 0\n",
    "    for l in range(len(merged_hsps)):\n",
    "        if(merged_hsps[l][4] > scoreMax):\n",
    "            scoreMax = merged_hsps[l][4]\n",
    "            best = l\n",
    "    \n",
    "    \n",
    "    return merged_hsps[best]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Alignement Smith-Waterman\n",
    "\n",
    "def traceback(startIndex, matrix, chemins, read1, read2):\n",
    "    \n",
    "    tracebackList1 = []\n",
    "    tracebackList2 = []\n",
    "    \n",
    "    i = startIndex[0]\n",
    "    j = startIndex[1]\n",
    "    \n",
    "    while(chemins[i,j] != -1):\n",
    "        if(chemins[i,j] == 0):\n",
    "            tracebackList1.append(\"-\")\n",
    "            tracebackList2.append(read2[i-1])\n",
    "            i = i-1\n",
    "        elif(chemins[i,j] == 1):\n",
    "            tracebackList1.append(read1[j-1])\n",
    "            tracebackList2.append(\"-\")\n",
    "            j = j-1\n",
    "        else:\n",
    "            tracebackList1.append(read1[j-1])\n",
    "            tracebackList2.append(read2[i-1])\n",
    "            i = i-1\n",
    "            j = j-1\n",
    "            \n",
    "    tracebackList1.reverse()\n",
    "    tracebackList2.reverse()\n",
    "    \n",
    "    return(tracebackList1, tracebackList2)\n",
    "\n",
    "def smithWaterman(seq1, seq2):\n",
    "    matrix = np.ones((len(seq2)+1, len(seq1)+1))\n",
    "    matrix[:,0] = 0\n",
    "    matrix[0] = 0\n",
    "\n",
    "    chemins = np.zeros((len(seq2)+1, len(seq1)+1))\n",
    "    chemins = chemins - 1\n",
    "\n",
    "    for i in range(1, len(seq2)+1):\n",
    "        for j in range(1, len(seq1)+1):\n",
    "            diagoValue = -1\n",
    "            if(seq1[j-1] == seq2[i-1]):\n",
    "                diagoValue = 1\n",
    "\n",
    "            matrix[i,j] = max(0, matrix[i-1,j] - 1, matrix[i,j-1] - 1, matrix[i-1,j-1] + diagoValue)\n",
    "            if(matrix[i,j] == 0):\n",
    "                chemins[i,j] = -1\n",
    "            elif(matrix[i,j] == matrix[i-1,j] - 1):\n",
    "                chemins[i,j] = 0\n",
    "            elif(matrix[i,j] == matrix[i,j-1] - 1):\n",
    "                chemins[i,j] = 1\n",
    "            else:\n",
    "                chemins[i,j] = 2\n",
    "    \n",
    "    scoreMax = matrix.max()\n",
    "    indice = np.unravel_index(np.argmax(matrix, axis=None), matrix.shape)\n",
    "                        \n",
    "    response = traceback(indice, matrix, chemins, seq1, seq2)\n",
    "    \n",
    "    # Pourcentage d'identité\n",
    "    match = 0\n",
    "    for i in range (len(response[0])):\n",
    "        if (response[0][i] == response[1][i]):\n",
    "            match += 1;\n",
    "    identity = match / len(response[0]) * 100\n",
    "    \n",
    "    return(scoreMax, response, len(response[0]), identity)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fonction pour transformer une liste en string\n",
    "\n",
    "def listToString(l):  \n",
    "    \n",
    "    string = \"\"  \n",
    "    for ele in l:  \n",
    "        string += ele  \n",
    "    \n",
    "    return string  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fonction de quicksort pour trier les résultats\n",
    "\n",
    "def sort(array):\n",
    "    less = []\n",
    "    equal = []\n",
    "    greater = []\n",
    "\n",
    "    if len(array) > 1:\n",
    "        pivot = array[0][0]\n",
    "        for x in array:\n",
    "            if x[0] < pivot:\n",
    "                less.append(x)\n",
    "            elif x[0] == pivot:\n",
    "                equal.append(x)\n",
    "            elif x[0] > pivot:\n",
    "                greater.append(x)\n",
    "        return sort(less)+equal+sort(greater) \n",
    "    else:  \n",
    "        return array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fonction pour exécuter pblast sur avec la seqNum-ième séquence du fichier sequences\n",
    "#la fonction utilise les valeurs par défaut\n",
    "\n",
    "def pblast(db=\"tRNAs.fasta\",sequences=\"unknown.fasta\",e=4,ss=0.001,seed=\"11111111111\",seqNum=1):\n",
    "    \n",
    "    #lignes de tRNAs.fasta\n",
    "    readsFile = open(db, \"r\")\n",
    "    readsLines = readsFile.readlines()\n",
    "    readsFile.close()\n",
    "\n",
    "    reads = []\n",
    "    for i in range(0, len(readsLines)):\n",
    "        reads.append(readsLines[i][:-1])\n",
    "\n",
    "    #lignes de unknown.fasta\n",
    "    readsFile2 = open(sequences, \"r\")\n",
    "    readsLines2 = readsFile2.readlines()\n",
    "    readsFile2.close()\n",
    "\n",
    "    reads2 = []\n",
    "    for i in range(0, len(readsLines2)):\n",
    "        reads2.append(readsLines2[i][:-1])\n",
    "\n",
    "    #assigner la séquence à blaster\n",
    "    sequence = reads2[seqNum*2-1]\n",
    "\n",
    "    #trouver le meilleur HSP pour chaque séquence de la BD\n",
    "    results = []\n",
    "    for i in range(0,len(reads),2):\n",
    "\n",
    "        #trouver le meilleur HSP\n",
    "        bestHsp = getBestHsp(sequence,reads[i+1],e,ss,seed)\n",
    "        if(bestHsp == None):\n",
    "            continue\n",
    "\n",
    "        #calculer les scores\n",
    "        bitscore = computeBitscore(bestHsp[4])\n",
    "        evalue = computeEvalue(len(reads[i+1]),len(sequence),bitscore)\n",
    "\n",
    "        #ne pas tenir compte du HSP s'il est au-dessus du 2e seuil\n",
    "        if(evalue > ss):\n",
    "            continue\n",
    "\n",
    "        #trouver l'alignement local\n",
    "        alignment = smithWaterman(sequence,reads[i+1])\n",
    "\n",
    "        resultString = \"\"\n",
    "        resultString += (\"%s  score: %.2f  ident: %.2f\\n\" % (reads[i],alignment[0],alignment[3]))\n",
    "        resultString += (listToString(alignment[1][1])+\"\\n\")\n",
    "        resultString += (listToString(alignment[1][0])+\"\\n\")\n",
    "        resultString += (\"\\n# Best HSP score: %.2f, bitscore: %.2f, evalue: %.2e\\n\" % (bestHsp[4],bitscore,evalue))\n",
    "        resultString += (\"%d  %s  %d\\n\" % (bestHsp[2], sequence[int(bestHsp[2]):int(bestHsp[3])+1], bestHsp[3]))\n",
    "        resultString += (\"%d  %s  %d\\n\\n\\n\" % (bestHsp[0], reads[i+1][int(bestHsp[0]):int(bestHsp[1])+1], bestHsp[1]))\n",
    "        resultString += (\"----------------------------------------\\n\")\n",
    "\n",
    "        results.append((alignment[3],resultString))\n",
    "\n",
    "    #trier les résultats et les afficher en ordre décroissant de % d'identité de l'alignement local\n",
    "    results = (sort(results))\n",
    "\n",
    "    out = \"\"\n",
    "    for j in range(len(results)-1,-1,-1):\n",
    "        out += results[j][1]\n",
    "    out += \"Total: \"+ str(len(results))\n",
    "    \n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">R|tct|Marchantia_polymorpha  score: 69.00  ident: 97.26\n",
      "GCATTCTTAGCTCAGTTGGATAGAGCAACAACCTTCTAAGTTGAAGGTCACAGGTTCAAATCCTGTAGAATGC\n",
      "GCATTCTTAGCTCAGCTGGATAGAGCAACAACCTTCTAAGTTGAAGGTCACAGGTTCAAATCCTGTAGGATGC\n",
      "\n",
      "# Best HSP score: 260.00, bitscore: 75.00, evalue: 1.43e-19\n",
      "16  TGGATAGAGCAACAACCTTCTAAGTTGAAGGTCACAGGTTCAAATCCTGTAG  67\n",
      "16  TGGATAGAGCAACAACCTTCTAAGTTGAAGGTCACAGGTTCAAATCCTGTAG  67\n",
      "\n",
      "\n",
      "----------------------------------------\n",
      ">R|tcg|Marchantia_polymorpha  score: 67.00  ident: 95.89\n",
      "GCATTCTTAGCTCAGTTGGATAGAGCAACAACCTTCGAAGTTGATGGTCACAGGTTCAAATCCTGTAGGATGC\n",
      "GCATTCTTAGCTCAGCTGGATAGAGCAACAACCTTCTAAGTTGAAGGTCACAGGTTCAAATCCTGTAGGATGC\n",
      "\n",
      "# Best HSP score: 140.00, bitscore: 41.00, evalue: 2.46e-09\n",
      "45  GGTCACAGGTTCAAATCCTGTAGGATGC  72\n",
      "45  GGTCACAGGTTCAAATCCTGTAGGATGC  72\n",
      "\n",
      "\n",
      "----------------------------------------\n",
      ">R|tct|Mesostigma_viride  score: 59.00  ident: 91.55\n",
      "CATTCTTAGCTCAGTTGGATAGAGCAACGGCCTTCTAAGCTGTAGGTCACAGGTTCAAATCCTGTAGAATG\n",
      "CATTCTTAGCTCAGCTGGATAGAGCAACAACCTTCTAAGTTGAAGGTCACAGGTTCAAATCCTGTAGGATG\n",
      "\n",
      "# Best HSP score: 120.00, bitscore: 36.00, evalue: 7.86e-08\n",
      "44  AGGTCACAGGTTCAAATCCTGTAG  67\n",
      "44  AGGTCACAGGTTCAAATCCTGTAG  67\n",
      "\n",
      "\n",
      "----------------------------------------\n",
      ">P|tgg|Marchantia_polymorpha  score: 26.00  ident: 71.67\n",
      "TAGCGCAATCTGG-TAGCGCGTCTGCCTTGGGCGCAGAAAGTTACAGGTTCAAATCCTGT\n",
      "TAGCTCAG-CTGGATAGAGCAACAACCTTCTAAGTTGAAGGTCACAGGTTCAAATCCTGT\n",
      "\n",
      "# Best HSP score: 85.00, bitscore: 26.00, evalue: 8.05e-05\n",
      "49  ACAGGTTCAAATCCTGT  65\n",
      "49  ACAGGTTCAAATCCTGT  65\n",
      "\n",
      "\n",
      "----------------------------------------\n",
      ">Y|gta|Bracteacoccus_minor  score: 22.00  ident: 68.97\n",
      "TGG-TGACGAGCTGTA-AACTCGT-TGACG--GAAG-TCTACGCAGGTTCAAATCCTG\n",
      "TGGAT-A-GAGCA--ACAAC-CTTCTAA-GTTGAAGGTC-AC--AGGTTCAAATCCTG\n",
      "\n",
      "# Best HSP score: 75.00, bitscore: 23.00, evalue: 7.22e-04\n",
      "50  CAGGTTCAAATCCTG  64\n",
      "59  CAGGTTCAAATCCTG  73\n",
      "\n",
      "\n",
      "----------------------------------------\n",
      ">P|tgg|Ourococcus_multisporus  score: 23.00  ident: 68.85\n",
      "TAGCGCAGCTTGG-TAGCGCAGTGGT-TT-TGGGAACCACAGGTCGCAGGTTCAAATCCTG\n",
      "TAGCTCAGCT-GGATAGAGCAACAACCTTCTAAGTTG-A-AGGTCACAGGTTCAAATCCTG\n",
      "\n",
      "# Best HSP score: 75.00, bitscore: 23.00, evalue: 6.44e-04\n",
      "50  CAGGTTCAAATCCTG  64\n",
      "50  CAGGTTCAAATCCTG  64\n",
      "\n",
      "\n",
      "----------------------------------------\n",
      ">Y|gta|Ourococcus_multisporus  score: 24.00  ident: 67.65\n",
      "GCTGAG-TGGTTAAAAGCG-CTAGACTGTAAATCTA-GT-GATATTGTCTACGCAGGTTCAAATCCTG\n",
      "GCTCAGCTGGATAGA-GCAAC-A-ACC-T---TCTAAGTTGA-AG-GTC-AC--AGGTTCAAATCCTG\n",
      "\n",
      "# Best HSP score: 75.00, bitscore: 23.00, evalue: 7.14e-04\n",
      "50  CAGGTTCAAATCCTG  64\n",
      "58  CAGGTTCAAATCCTG  72\n",
      "\n",
      "\n",
      "----------------------------------------\n",
      ">Y|gta|Bracteacoccus_aerius  score: 24.00  ident: 67.65\n",
      "AGCGGTCAA-TGG-TGACGAGCTGTA-AACTCGT-TGACG--GAAG-TCTACGCAGGTTCAAATCCTG\n",
      "AGC--TCAGCTGGAT-A-GAGCA--ACAAC-CTTCTAA-GTTGAAGGTC-AC--AGGTTCAAATCCTG\n",
      "\n",
      "# Best HSP score: 75.00, bitscore: 23.00, evalue: 7.22e-04\n",
      "50  CAGGTTCAAATCCTG  64\n",
      "59  CAGGTTCAAATCCTG  73\n",
      "\n",
      "\n",
      "----------------------------------------\n",
      ">M|cat|Neochloris_aquatica  score: 21.00  ident: 66.67\n",
      "TAGAGCAATTGGTTAG--CTTATCGGGCT-CATAA-TCCGACGGT-ATAGGTTCAAATCCTGT\n",
      "TAGCTCAGCTGGATAGAGCA-A-CAACCTTC-TAAGTT-GAAGGTCACAGGTTCAAATCCTGT\n",
      "\n",
      "# Best HSP score: 75.00, bitscore: 23.00, evalue: 6.35e-04\n",
      "51  AGGTTCAAATCCTGT  65\n",
      "50  AGGTTCAAATCCTGT  64\n",
      "\n",
      "\n",
      "----------------------------------------\n",
      ">P|tgg|Mesostigma_viride  score: 21.00  ident: 66.67\n",
      "TAACGCAGCTAGG-TAGCGTGTCTGT-TT-TGGGAACA-GAATGCCACAGGTTCAAATCCTGT\n",
      "TAGCTCAGCT-GGATAGAGCAACAACCTTCT---AAGTTGAAGGTCACAGGTTCAAATCCTGT\n",
      "\n",
      "# Best HSP score: 90.00, bitscore: 27.00, evalue: 4.02e-05\n",
      "48  CACAGGTTCAAATCCTGT  65\n",
      "48  CACAGGTTCAAATCCTGT  65\n",
      "\n",
      "\n",
      "----------------------------------------\n",
      ">S|gga|Neochloris_aquatica  score: 22.00  ident: 66.18\n",
      "CAGAGTGGTTTA-ATGCGCCAAC-TTGGAAGGTTGGTTCCCTTTAGGGATCACAGGTTCAAATCCTGT\n",
      "CAGC-TGGAT-AGA-GCAACAACCTTCTAAG-TTGA--------AGG--TCACAGGTTCAAATCCTGT\n",
      "\n",
      "# Best HSP score: 95.00, bitscore: 29.00, evalue: 1.14e-05\n",
      "47  TCACAGGTTCAAATCCTGT  65\n",
      "57  TCACAGGTTCAAATCCTGT  75\n",
      "\n",
      "\n",
      "----------------------------------------\n",
      "Total: 11\n"
     ]
    }
   ],
   "source": [
    "#Permet d'appeler pBLAST avec différents paramètre\n",
    "#1.On n'affiche les résultat pour une seule séquence du fichier \"unknown.fasta\" à la fois. Celui-ci\n",
    "#doit être décidé avec l'argument \"seqNum\" (1 à 4)\n",
    "#2.Les paramètre pouvant être changés sont:\n",
    "# - db: le fichier de base de données (doit être dans le même dossier que le code)\n",
    "# - sequences: le fichier de séquences à rechercher (doit être dans le même dossier que le code)\n",
    "# - e: le 1er seuil\n",
    "# - ss: le 2e seuil\n",
    "# - seed: la graine (composée de 1 et de 0)\n",
    "# - seqNum: le choix de la séquence à blaster\n",
    "\n",
    "print(pblast(seqNum=4))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ift3700",
   "language": "python",
   "name": "ift3700"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
